src/main/java/hudson/remoting/RemoteInvocationHandler.java

```diff
@@ -233,16 +292,217 @@ public int hashCode() {
         return oid;
     }
 
+    /**
+     * Finalizers are run only under extreme GC pressure whereas {@link PhantomReference} are cleared out
+     * more quickly, thus we use a {@link PhantomReference} in place of the override of {@link Object#finalize()}
+     * that was previously used in order to unexport.
+     * @since FIXME after merge
+     */
+    private static class PhantomReferenceImpl extends PhantomReference<RemoteInvocationHandler> {
 
-    protected void finalize() throws Throwable {
-        // unexport the remote object
-        if (channel!=null && !autoUnexportByCaller) {
-            channel.send(new UnexportCommand(oid,origin));
-            channel = null;
+        /**
+         * The object id to unexport.
+         */
+        private final int oid;
+        /**
+         * The origin from where the object was created.
+         */
+        private Throwable origin;
+        /**
+         * The reference to the channel on which to unexport.
+         */
+        private Channel.Ref channel;
+
+        /**
+         * Construct our reference and bind to the {@link ReferenceQueue} after capturing the required state for
+         * {@link #cleanup()}.
+         *
+         * @param referent       the {@link RemoteInvocationHandler} we will clean up.
+         * @param referenceQueue the {@link ReferenceQueue}
+         */
+        private PhantomReferenceImpl(RemoteInvocationHandler referent,
+                                    ReferenceQueue<? super RemoteInvocationHandler> referenceQueue) {
+            super(referent, referenceQueue);
+            this.oid = referent.oid;
+            this.origin = referent.origin;
+            this.channel = referent.channel;
+        }
+
+        /**
+         * Sends the {@link UnexportCommand} for the specified {@link #oid} if the {@link Channel} is still open.
+         * @throws IOException if the {@link UnexportCommand} could not be sent.
+         */
+        private void cleanup() throws IOException {
+            if (this.channel == null) {
+                return;
+            }
+            Channel channel = this.channel.channel();
+            if (channel != null && !channel.isClosingOrClosed()) {
+                try {
+                    channel.send(new UnexportCommand(oid, origin));
+                } finally {
+                    // clear out references to simplify GC
+                    this.origin = null;
+                    this.channel = null;
+                }
+            }
         }
-        super.finalize();
     }
 
+    /**
+     * Manages the cleanup of {@link RemoteInvocationHandler} instances that need to be auto unexported.
+     * @since FIXME after merge
+     */
+    private static class Unexporter implements Runnable {
+
+        /**
+         * Our executor service, we use at most one thread for all {@link Channel} instances in the current classloader.
+         */
+        private final ExecutorService svc = new AtmostOneThreadExecutor(
```

Comment:
No need because `UNEXPORTER` is static. This leaves us with just the one static to track